xds/src/test/java/io/grpc/xds/EdsLoadBalancer2Test.java

```diff
@@ -0,0 +1,836 @@
+    assertThat(config.targets).hasSize(2);
+    WeightedPolicySelection target1 = config.targets.get(locality1.toString());
+    assertThat(target1.weight).isEqualTo(30);
+    assertThat(target1.policySelection.getProvider().getPolicyName()).isEqualTo(LRS_POLICY_NAME);
+    assertLrsConfig((LrsConfig) target1.policySelection.getConfig(), CLUSTER, EDS_SERVICE_NAME,
+        LRS_SERVER_NAME, locality1, "round_robin");
+
+    WeightedPolicySelection target2 = config.targets.get(locality2.toString());
+    assertThat(target2.weight).isEqualTo(40);
+    assertThat(target2.policySelection.getProvider().getPolicyName()).isEqualTo(LRS_POLICY_NAME);
+    assertLrsConfig((LrsConfig) target2.policySelection.getConfig(), CLUSTER, EDS_SERVICE_NAME,
+        LRS_SERVER_NAME, locality2, "round_robin");
+  }
+
+  @Test
+  public void generateWeightedTargetLbConfig_withoutLrsPolicy() {
+    Map<Locality, Integer> localityWeights = ImmutableMap.of(locality1, 30, locality2, 40);
+    WeightedTargetConfig config =
+        EdsLoadBalancer2.generateWeightedTargetLbConfig(
+            CLUSTER, EDS_SERVICE_NAME, null, roundRobin, registry, localityWeights);
+    assertThat(config.targets).hasSize(2);
+    WeightedPolicySelection target1 = config.targets.get(locality1.toString());
+    assertThat(target1.weight).isEqualTo(30);
+    assertThat(target1.policySelection.getProvider().getPolicyName()).isEqualTo("round_robin");
+
+    WeightedPolicySelection target2 = config.targets.get(locality2.toString());
+    assertThat(target2.weight).isEqualTo(40);
+    assertThat(target2.policySelection.getProvider().getPolicyName()).isEqualTo("round_robin");
+  }
+
+  private void deliverSimpleClusterLoadAssignment(String resourceName) {
+    EquivalentAddressGroup endpoint1 = makeAddress("endpoint-addr-1");
+    LocalityLbEndpoints localityLbEndpoints1 =
+        buildLocalityLbEndpoints(1, 10, Collections.singletonList(endpoint1));
+    xdsClient.deliverClusterLoadAssignment(
+        resourceName,
+        Collections.singletonMap(locality1, localityLbEndpoints1));
+  }
+
+  /**
+   * Recreate a new {@link EdsLoadBalancer2} that will instantiate its downstream LB policies with
+   * real implementations, except the leaf policy is replaced with a fake implementation to avoid
+   * creating connections.
+   */
+  private void recreateLoadBalancerWithRealDownstreamLbPolicies(
+      FakeLoadBalancerProvider fakeLeafPolicyProvider) {
+    LoadBalancerRegistry lbRegistry = LoadBalancerRegistry.getDefaultRegistry();
+    lbRegistry.deregister(lbRegistry.getProvider(fakeLeafPolicyProvider.getPolicyName()));
+    lbRegistry.register(fakeLeafPolicyProvider);
+    loadBalancer.shutdown();
+    loadBalancer = new EdsLoadBalancer2(helper, lbRegistry, mockRandom);
+    PolicySelection weightedTargetSelection =
+        new PolicySelection(lbRegistry.getProvider(WEIGHTED_TARGET_POLICY_NAME), null, null);
```

Comment:
An alternative is: deregister all the _fake_ providers in the `setUp()`, and re-register those policies with _real_ providers. Then the benefit is you don't need to shutdown the balancer and create a new balancer.
```java
    registry.deregister(registry.getProvider(PRIORITY_POLICY_NAME));
    registry.register(new PriorityLoadBalancerProvider());
    registry.deregister(registry.getProvider(LRS_POLICY_NAME));
    registry.register(new LrsLoadBalancerProvider());
    registry.register(fakeLeafPolicyProvider);
    PolicySelection weightedTargetSelection =
        new PolicySelection(new WeightedTargetLoadBalancerProvider(), null, null);
```

Reply:
If using real registry, you need clean up the real registry at the end of test.

Reply:
Fixed to reuse the existing registry with replaced LB providers.

Reply:
Hmm... It turns out even without
```java
lbRegistry.register(fakeLeafPolicyProvider);
```
the test could still pass.

Reply:
Because the leaf policy provider is not obtained from the registry, it is obtained from the LB config. So the one in the registry is not used.